Runtime/IJSValue.cs

```diff
@@ -0,0 +1,23 @@
+using System;
+
+namespace NodeApi;
+
+/// <summary>
+/// Interface for a type that is convertible to and equatable to a JS value.
+/// </summary>
+/// <remarks>
+/// This interface is implemented by:
+///   - Value types that directly represent a specific type of JS value such as
+///     <see cref="JSObject" />, <see cref="JSArray" />, etc.
+///   - Collection proxy classes returned by <see cref="JSIterable.AsEnumerable()" /> and similar
+///     methods, that forward all calls to a JS collection object.
+///   - Interface proxy classes auto-generated by <see cref="JSExportAttribute" />, that forward
+///     all calls to a JS object that implements the interface.
+/// </remarks>
+public interface IJSValue : IEquatable<JSValue>
```

Comment:
>IJSValue [](http://example.com/codeflow?start=17&length=8)

It means that now we must box and unbox JSValues. It goes against us being efficient and performant. Could we use an alternative solution with static abstract methods instead that would help us avoid boxing?

Reply:
> It means that now we must box and unbox JSValues

No, not normally.

`IJSValue` is meant to be used when you have some _already boxed_ object and you need to find out whether it represents a JS value. It doesn't change any of the existing patterns that avoid boxing. Does that make sense?

Reply:
I guess it makes sense then. I am not sure if the `IJSValue` is the best name for an interface for such purpose.

In one of my prototypes, I implemented [`ITypedValue`](https://github.com/vmoroz/napi-dotnet/blob/dev/vmoroz/prototypes/typed_model/Runtime/TypedModel/TypedModel.cs) that used to get `JSValue`. In that code the interface and custom attributes were used to generate boilerplate code for the partial structs and use them uniformly. It was a static interface to avoid boxing/unboxing. We could use the name `IJSValue` instead of `ITypedValue` and do the same.

Another concern is that this interface looks so "basic" that other developers may be tempted to use it and think that it is a good practice. E.g. to write extensions methods. I would rather have them writing extensions based on the static interfaces instead of the boxed value. I wonder if the marshaller can be smart enough to look for structs and then see if it has a static interface implemented.

Reply:
A static interface won't work for the same purpose because the generic type argument is required while the precise type of the object isn't known.

I see your point about how the interface can be confusing. I can remove it and use a different approach in the marshaller that doesn't rely on a public interface.

Reply:
Removed. It wasn't really needed, the code is simpler now.